src/main/scala/beam/utils/scenario/urbansim/HOVModeTransformer.scala

```diff
@@ -0,0 +1,419 @@
+    persons: Iterable[PersonInfo],
+    households: Iterable[HouseholdInfo]
+  ): Iterable[PlanElement] = {
+    val householdsCarCount: mutable.HashMap[HouseholdId, Int] =
+      mutable.HashMap(households.map(hh => hh.householdId -> hh.cars).toSeq: _*)
+
+    val allHOVUsers: Set[PersonId] = plansProbablyWithHOV
+      .filter(planElement => {
+        val legMode = planElement.legMode.map(_.toLowerCase)
+        legMode.contains(hov2) || legMode.contains(hov3)
+      })
+      .map(_.personId)
+      .toSet
+
+    val allExpectedCarUsers: Set[PersonId] = plansProbablyWithHOV
+      .filter(
+        planElement => planElement.legMode.exists(mode => allCarModesWithHOV.contains(mode.toLowerCase))
+      )
+      .map(_.personId)
+      .toSet
+
+    val didNotGetACar = mutable.Set.empty[PersonId]
+    val didGetACar = mutable.Set.empty[PersonId]
+    persons.foreach { person =>
+      if (!didGetACar.contains(person.personId) && allExpectedCarUsers.contains(person.personId)) {
+        householdsCarCount.get(person.householdId) match {
+          case Some(value) if value > 0 =>
+            didGetACar.add(person.personId)
+            householdsCarCount(person.householdId) = value - 1
+          case _ => didNotGetACar.add(person.personId)
+        }
+      }
+    }
+
+    var forcedHOV2Teleports = 0
+    var forcedHOV3Teleports = 0
+
+    val plansWithAccessToCar = if (didNotGetACar.nonEmpty) {
+      plansProbablyWithHOV.map { planElement =>
+        val personDidNotGetACar = didNotGetACar.contains(planElement.personId)
+
+        planElement.legMode match {
+          case Some(legMode) if personDidNotGetACar && isHOV2(legMode) =>
+            forcedHOV2Teleports += 1
+            planElement.copy(legMode = Some(HOV2_TELEPORTATION.value))
+
+          case Some(legMode) if personDidNotGetACar && isHOV3(legMode) =>
+            forcedHOV3Teleports += 1
+            planElement.copy(legMode = Some(HOV3_TELEPORTATION.value))
+
+          case _ => planElement
+        }
+      }
+    } else {
+      plansProbablyWithHOV
+    }
+
+    if (forcedHOV2Teleports > 0 || forcedHOV3Teleports > 0) {
+      logger.info(
+        s"There were $forcedHOV2Teleports hov2 and $forcedHOV3Teleports hov3 forced teleports because actors did not get access to a car."
+      )
+    }
+
+    val withCarAccessHOVUsers = allHOVUsers -- didNotGetACar
+
+    var forcedCarHOV2Count = 0
+    var forcedCarHOV3Count = 0
+
+    def thereAreMoreHOVTeleportations: Boolean = {
+      val forcedTeleportsCount = forcedHOV2Teleports + forcedHOV3Teleports
+      forcedTeleportsCount > 0 && forcedTeleportsCount > forcedCarHOV2Count + forcedCarHOV3Count
+    }
+
+    def thereAreMoreHOVCars: Boolean = {
+      val forcedCarHOVCount = forcedCarHOV2Count + forcedCarHOV3Count
+      forcedCarHOVCount > 0 && forcedCarHOVCount > forcedHOV2Teleports + forcedHOV3Teleports
+    }
+
+    def mapToHovCarBalancingForcedTeleports(trip: Iterable[PlanElement]): Iterable[PlanElement] = {
```

Comment:
names of these two functions are confusing, 
I assume something like `raplaceHOVwithCar` and 'replaceHOVwithTeleportation'